distributed/tests/test_stories.py

```diff
@@ -0,0 +1,170 @@
+        [
+            (key, "released", "waiting", {key: "processing"}),
+            (key, "waiting", "processing", {}),
+            (key, "processing", "memory", {}),
+        ],
+    )
+
+    await c.close()
+
+    stories_after_close = s.story(key)
+    assert len(stories_after_close) > len(stories)
+
+    stimulus_ids = {s[-2] for s in stories_after_close}
+    # One more event
+    # - Forget / Release / Free since client closed
+    assert len(stimulus_ids) == 3
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_scheduler_story_stimulus_retry(c, s, a):
+    def task():
+        assert dask.config.get("foo")
+
+    with dask.config.set(foo=False):
+        f = c.submit(task)
+        with pytest.raises(AssertionError):
+            await f
+
+    with dask.config.set(foo=True):
+        await f.retry()
+        await f
+
+    story = s.story(f.key)
+    stimulus_ids = {s[-2] for s in story}
+    # Four events
+    # - Compute
+    # - Erred
+    # - Compute / Retry
+    # - Success
+    assert len(stimulus_ids) == 4
+
+    assert_story(
+        story,
+        [
+            # Erred transitions via released
+            (f.key, "processing", "erred", {}),
+            (f.key, "erred", "released", {}),
+            (f.key, "released", "waiting", {f.key: "processing"}),
+        ],
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_client_story(c, s, a):
+    f = c.submit(inc, 1)
+    assert await f == 2
+    story = await c.story(f.key)
+
+    # Every event should be prefixed with it's origin
+    # This changes the format compared to default scheduler / worker stories
+    prefixes = set()
+    stripped_story = list()
+    for msg in story:
+        prefixes.add(msg[0])
+        stripped_story.append(msg[1:])
+    assert prefixes == {"scheduler", a.address}
+
+    assert_valid_story(stripped_story, ordered_timestamps=False)
+
+    # If it's a well formed story, we can sort by the last element which is a
+    # timestamp and compare the two lists.
+    assert sorted(stripped_story, key=lambda msg: msg[-1]) == sorted(
+        s.story(f.key) + a.story(f.key), key=lambda msg: msg[-1]
+    )
+
+
+class WorkerBrokenStory(Worker):
+    async def get_story(self, *args, **kw):
+        raise CommClosedError
+
+
+@gen_cluster(client=True, Worker=WorkerBrokenStory)
+@pytest.mark.parametrize("on_error", ["ignore", "raise"])
+async def test_client_story_failed_worker(c, s, a, b, on_error):
+    f = c.submit(inc, 1)
+    coro = c.story(f.key, on_error=on_error)
+    await f
+
+    if on_error == "raise":
+        with pytest.raises(CommClosedError):
+            await coro
+    elif on_error == "ignore":
+        story = await coro
+        assert story
+        assert len(story) > 1
+    else:
+        raise ValueError(on_error)
+
+
+@gen_cluster(client=True)
+async def test_worker_story_with_deps(c, s, a, b):
+    """
+    Assert that the structure of the story does not change unintentionally and
```

Comment:
I'm not sure I agree with this as a motivation for a test.  We should be ok with things changing.  We should be excited about things changing.  However, if there are particular changes that would concern us, then It hink that we should focus on and make sure that we're testing those.  This is a good way to communicate to future developers what is ok to change and what is not ok to change.  If we just say "fail if anything changes" then we're asking future developers to make this decision, and they may be less able than we are today.

Reply:
Oh, or if you're considering this as public API, then maybe it would make sense to test the structure of a story?  Tuples of a certain length, etc..  I would be totally ok with that.

Reply:
I left an answer to this in the other thread